covid_graphs/covid_graphs/country_report.py

```diff
@@ -1,29 +1,52 @@
 import datetime
+from dataclasses import dataclass
 from pathlib import Path
+from typing import List
 
 import numpy as np
 from google.protobuf import text_format  # type: ignore
 
 from .pb.country_data_pb2 import CountryData
 
 
+@dataclass
 class CountryReport:
-    def __init__(self, country_data_file: Path, short_name: str):
-        """Constructs a numpy representation of data read from 'data_dir' a given country."""
-        country_data = CountryData()
-        text_format.Parse(country_data_file.read_text(), country_data)
-
-        self.short_name = short_name
-        self.long_name = country_data.name
-        self.dates = [
-            datetime.date(day=day.date.day, month=day.date.month, year=day.date.year)
-            for day in country_data.stats
-        ]
-        self.dates_str = list(map(lambda d: d.strftime("%Y-%m-%d"), self.dates))
-
-        self.daily_positive = np.array([day.positive for day in country_data.stats])
-        self.daily_dead = np.array([day.dead for day in country_data.stats])
-        self.daily_recovered = np.array([day.recovered for day in country_data.stats])
-        self.daily_active = self.daily_positive - self.daily_recovered - self.daily_dead
-
-        self.cumulative_active = np.add.accumulate(self.daily_active)
+    short_name: str
+    long_name: str
+    dates: List[datetime.date]
+    daily_positive: np.ndarray
+    daily_dead: np.ndarray
+    daily_recovered: np.ndarray
+    daily_active: np.ndarray
+    cumulative_active: np.ndarray
+
+
+def create_report(country_data_file: Path, short_name: str) -> CountryReport:
+    """Constructs a numpy representation of data read from 'data_dir' a given country."""
+    country_data = CountryData()
+    text_format.Parse(country_data_file.read_text(), country_data)
+
+    short_name = short_name
+    long_name = country_data.name
+    dates = [
+        datetime.date(day=day.date.day, month=day.date.month, year=day.date.year)
+        for day in country_data.stats
+    ]
+
+    daily_positive = np.array([day.positive for day in country_data.stats])
+    daily_dead = np.array([day.dead for day in country_data.stats])
+    daily_recovered = np.array([day.recovered for day in country_data.stats])
+    daily_active = daily_positive - daily_recovered - daily_dead
+
+    cumulative_active = np.add.accumulate(daily_active)
+
+    return CountryReport(
+        short_name,
+        long_name,
+        dates,
+        daily_positive,
+        daily_dead,
+        daily_recovered,
+        daily_active,
+        cumulative_active,
+    )
```

Comment:
I prefer named calls (`short_name=short_name`) - it is in my opinion clearer and it prevents you from calls with wrong parameters when you change the function signature. In this case it feels a little silly though; it is more about style preference, so I'll leave it to you.

Reply:
Coming from the C++ world, I'm used to the Google C++ style, whose philosophy can be applied here as well.

When a function is called with an argument that doesn't have a name (i.e. a constant) or has a different name, then the original argument name is written as a comment ![Screenshot_20200425_072919](https://user-images.githubusercontent.com/3822689/80272086-b8c84280-86c6-11ea-8476-be2662e93cb3.png)

In Python, that would be equivalent to `CalculateProduct(values, options, completion_callback=None)`. It seems to me to be the most reasonable approach. I think that writing `CalculateProduct(values=values, options=options, completion_callback=None)` **hurts** readability, since you are adding redundant information that takes up space.

Reply:
Unfortunately readability is a function of convention, and what is worse that differs for every language... But I'm fine with the way you kept it, especially for a small project with a couple of devs like this I wouldn't be too fussy.